File: 0113-path-sum-ii/solution.java
class Solution {
    public List<List<Integer>> pathSum(TreeNode root, int targetSum) {
        List<List<Integer>> result = new ArrayList<>();
        List<Integer> path = new ArrayList<>();
        construct(root, path, result, targetSum);
        return result;
    }

    private void construct(TreeNode node, List<Integer> path, List<List<Integer>> result, int target) {
        if (node == null) return;

        path.add(node.val);
        target -= node.val;
        if (node.left == null && node.right == null && target == 0) {
            result.add(new ArrayList<>(path));
        }

        construct(node.left, path, result, target);
        construct(node.right, path, result, target);
        path.remove(path.size() - 1);
    }
}

File: 0113-path-sum-ii/solution.cpp
class Solution {
public:
    void helper(TreeNode* node, vector<vector<int>> &result, vector<int> &path, int target) {
        if (node == nullptr) return;
        path.push_back(node->val);
        target -= node->val;
        if (node->left == nullptr && node->right == nullptr && target == 0) {
            result.push_back(path);
        }
        helper(node->left, result, path, target);
        helper(node->right, result, path, target);
        path.pop_back();
    }
    vector<vector<int>> pathSum(TreeNode* root, int targetSum) {
        vector<vector<int>> result;
        vector<int> path;
        helper(root, result, path, targetSum);
        return result;
    }
};

